File: ts/src/session.ts
import type { Event, SessionInfo } from "./types.js";
import { HttpError } from "./errors.js";
import { parseSSE } from "./sse.js";

type FetchFn = typeof globalThis.fetch;

export class Session {
  readonly #baseURL: string;
  readonly #fetch: FetchFn;
  readonly #info: SessionInfo;
  #lastAckedSeq: number = 0;

  /** @internal Created by Client.createSession, not for direct use. */
  constructor(baseURL: string, fetchFn: FetchFn, info: SessionInfo) {
    this.#baseURL = baseURL;
    this.#fetch = fetchFn;
    this.#info = info;
  }

  get info(): SessionInfo {
    return this.#info;
  }

  get lastAckedSeq(): number {
    return this.#lastAckedSeq;
  }

  /**
   * Open the SSE stream for this session. Replays buffered frames
   * from the cursor, then streams live.
   */
  async subscribe(signal?: AbortSignal): Promise<AsyncIterable<Event>> {
    const url = `${this.#baseURL}/clients/${this.#info.client_id}/events`;
    const resp = await this.#fetch(url, {
      headers: { Accept: "text/event-stream" },
      signal,
    });

    if (!resp.ok) {
      const body = await resp.text();
      throw new HttpError("GET", url, resp.status, body);
    }

    if (!resp.body) {
      throw new HttpError("GET", url, resp.status, "no response body");
    }

    return parseSSE(resp.body);
  }

  /** Advance the cursor for this session to the given sequence number. */
  async ack(seq: number, signal?: AbortSignal): Promise<void> {
    const url = `${this.#baseURL}/clients/${this.#info.client_id}/ack`;
    const resp = await this.#fetch(url, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ seq }),
      signal,
    });

    if (resp.status !== 204) {
      const body = await resp.text();
      throw new HttpError("PUT", url, resp.status, body);
    }

    this.#lastAckedSeq = seq;
  }
}

File: lplexc/session.go
package lplexc

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// SessionConfig configures a buffered client session.
type SessionConfig struct {
	ClientID      string
	BufferTimeout string // ISO 8601 duration, e.g. "PT5M"
	Filter        *Filter
	AckInterval   time.Duration // how often to auto-ACK (0 = manual only)
}

// SessionInfo is the response from creating or reconnecting a session.
type SessionInfo struct {
	ClientID string   `json:"client_id"`
	Seq      uint64   `json:"seq"`
	Cursor   uint64   `json:"cursor"`
	Devices  []Device `json:"devices"`
}

// Session is a buffered connection to an lplex server with cursor tracking
// and automatic reconnect replay.
type Session struct {
	client    *Client
	config    SessionConfig
	info      SessionInfo
	lastAcked uint64
}

// CreateSession creates or reconnects a buffered session on the server.
func (c *Client) CreateSession(ctx context.Context, cfg SessionConfig) (*Session, error) {
	sessionURL := c.baseURL + "/clients/" + cfg.ClientID

	putBody := map[string]any{"buffer_timeout": cfg.BufferTimeout}
	if cfg.Filter != nil && !cfg.Filter.IsEmpty() {
		putBody["filter"] = filterSessionJSON(cfg.Filter)
	}
	body, _ := json.Marshal(putBody)

	req, err := http.NewRequestWithContext(ctx, http.MethodPut, sessionURL, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		errBody, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("PUT /clients/%s returned %d: %s", cfg.ClientID, resp.StatusCode, errBody)
	}

	var info SessionInfo
	if err := json.NewDecoder(resp.Body).Decode(&info); err != nil {
		return nil, fmt.Errorf("decoding session: %w", err)
	}

	return &Session{
		client: c,
		config: cfg,
		info:   info,
	}, nil
}

// Info returns the session metadata from the server.
func (s *Session) Info() SessionInfo {
	return s.info
}

// Subscribe opens the SSE stream for this session (GET /clients/{id}/events).
// Replays any buffered frames from the cursor, then streams live.
func (s *Session) Subscribe(ctx context.Context) (*Subscription, error) {
	eventsURL := s.client.baseURL + "/clients/" + s.config.ClientID + "/events"

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, eventsURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "text/event-stream")

	resp, err := s.client.httpClient.Do(req)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		_ = resp.Body.Close()
		return nil, fmt.Errorf("GET /clients/%s/events returned %d: %s", s.config.ClientID, resp.StatusCode, body)
	}

	return newSubscription(resp.Body), nil
}

// Ack advances the cursor for this session to the given sequence number.
func (s *Session) Ack(ctx context.Context, seq uint64) error {
	ackURL := s.client.baseURL + "/clients/" + s.config.ClientID + "/ack"
	body, _ := json.Marshal(map[string]uint64{"seq": seq})

	req, err := http.NewRequestWithContext(ctx, http.MethodPut, ackURL, strings.NewReader(string(body)))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := s.client.httpClient.Do(req)
	if err != nil {
		return err
	}
	_ = resp.Body.Close()

	if resp.StatusCode != http.StatusNoContent {
		return fmt.Errorf("ack returned %d", resp.StatusCode)
	}
	s.lastAcked = seq
	return nil
}

// LastAcked returns the last sequence number that was successfully acknowledged.
func (s *Session) LastAcked() uint64 {
	return s.lastAcked
}

// filterSessionJSON converts a Filter to the JSON body format used by PUT /clients/{id}.
func filterSessionJSON(f *Filter) map[string]any {
	m := map[string]any{}
	if len(f.PGNs) > 0 {
		m["pgn"] = f.PGNs
	}
	if len(f.ExcludePGNs) > 0 {
		m["exclude_pgn"] = f.ExcludePGNs
	}
	if len(f.Manufacturers) > 0 {
		m["manufacturer"] = f.Manufacturers
	}
	if len(f.Instances) > 0 {
		m["instance"] = f.Instances
	}
	if len(f.Names) > 0 {
		m["name"] = f.Names
	}
	if len(f.ExcludeNames) > 0 {
		m["exclude_name"] = f.ExcludeNames
	}
	return m
}
